File: domino-rotations.java
// Time , Space - O(N), O(N)
/**
 * @param {number[]} A
 * @param {number[]} B
 * @return {number}
 */
var minDominoRotations = function(A, B) {
    if(!A || A.length == 0) {
        return 0;
    }
    let map = {};
    let maxFreq = 0, max = -1;
    
    for(let i=0;i<A.length;i++) {
        map = {
            ...map,
            [A[i]]: map[A[i]] ? map[A[i]] + 1 : 1
        }
        maxFreq = Math.max(maxFreq, map[A[i]]);
        if(maxFreq >= A.length) {
            max = A[i];
            break;
        }
        
        map = {
            ...map,
            [B[i]]: map[B[i]] ? map[B[i]] + 1 : 1
        }
        if(maxFreq >= A.length) {
            max = B[i];
            break;
        }
        
    }
    if(max == -1) {
        return max;
    }
    let aRot = 0, bRot = 0;
    for(let i=0;i<A.length;i++) {
        if(A[i]!=max &&B[i]!=max) {
            return -1;
        }
        if(A[i] != max) {
            aRot++;
        }
        if(B[i] != max) {
            bRot++;
        }
    }
    
    return Math.min(aRot, bRot);
    
    
};

File: DominoRotations.py
# Time Complexity : O(n) 
# Space Complexity :O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# class Solution:
#     def minDominoRotations(self, A, B):
#         maxFreq = 0 
#         for i in range(1, 7):
#             count = 0 
#             for j in range(len(A)):
#                 if A[j] == i :
#                     count +=1 
#                 if B[j] == i:
#                     count += 1 
#             if count >= len(A):
#                 maxFreq = i

#         maxA = 0 
#         maxB = 0 
#         for i in range(len(A)):
#             if A[i] != maxFreq and B[i] != maxFreq: 
#                 return -1 
#             if A[i] != maxFreq:
#                 maxA += 1 
            
#             if B[i] != maxFreq:
#                 maxB += 1 
        
#         return min(maxA, maxB)

class Solution:
    def minDominoRotations(self, A, B):
        if len(A) == 0 or A is None :
            return 0
        res = self.check(A, B, B[0])
        if res != -1:
            return res 
        return self.check(A, B, A[0])
    
    def check(self, A, B, target):
        maxA = 0 
        maxB = 0 
        for i in range(len(A)):
            if A[i] != target and B[i] != target: 
                return -1 
            if A[i] != target:
                maxA += 1 
            
            if B[i] != target:
                maxB += 1 
        
        return min(maxA, maxB)
    
if __name__ == "__main__":
    s = Solution()
    A = [2,1,2,4,2,2]
    B = [5,2,6,2,3,2]
    res = s.minDominoRotations(A, B )
    assert res == 2
        